File: reportBuilder.py
# -*- coding: utf-8 -*-
"""
Created on Sun Oct 21 17:41:59 2018

@author: mikeh
"""

import pandas 
import seaborn as sns
import os
from datetime import date
from dateutil.relativedelta import relativedelta as rd


def makeDirs (indir="C:/Users/mikeh/Desktop/Projects/bd_reportBuilder/data", outdir="C:/Users/mikeh/Desktop/Projects/bd_reportBuilder/reports"):
    
    if not os.path.exists(indir):           
        os.makedirs(indir)
    os.chdir(indir)
   
    if not os.path.exists(outdir):           
        os.makedirs(outdir)
    
    print('Directories established')
    

def inputValidation ():

    monthInRange = False
    isAYear = False
    
    firmName = input("Enter the Firm Name: ")
    firmName = firmName.strip()
    
    """
    Firm name will be selected via a drop down called from db generated firm list
    
    """
    
    while not monthInRange:
        endMonth = input("Please enter a month by number, eg. for January use 1, December 12: ")
        endMonth = endMonth.strip()
        endMonth = int(endMonth)
                    
        if endMonth > 0 and endMonth < 13:
            monthInRange = True
    
    while not isAYear:
        endYear = input("Please enter a year, eg. 2018: ")
        endYear = endYear.strip()        
        if len(endYear) == 4:
            endYear = int(endYear)
            isAYear = True    
    
    print("To confirm the following will be calculated:")
    print("Firm:", firmName)
    print("End Month:", endMonth)
    print("End Year:", endYear)
    
    
    
    
def MakeRelevantDates (endMonth, endYear, yearEnd=12):
    
    endOfMonth = (31,28,31,30,31,30,31,31,30,31,30,31)
    leapYearEndOfMonth = (31,29,31,30,31,30,31,31,30,31,30,31)

    
    if endYear % 4 == 0:
        endDate = date(endYear, endMonth, leapYearEndOfMonth[endMonth-1])
        endDatePrior = endDate + rd(years=-1)
    elif (endYear-1) % 4 == 0:
        endDatePrior = date(endYear - 1, endMonth, leapYearEndOfMonth[endMonth-1])
        endDate = endDatePrior + rd(years=+1)
    else:
        endDate = date(endYear, endMonth, endOfMonth[endMonth-1])
        endDatePrior = endDate + rd(years=-1)
    
    if endMonth == 12:
        startMonth = 1
        startYear = endYear
    else:
        startMonth = endMonth + 1
        startYear = endYear - 1
        
    startDate = date(startYear, startMonth, 1)
    startDatePrior = startDate + rd(years=-1)
    
    if endMonth == yearEnd:
        fiscalYE = endDate
        fiscalYEm1 = endDatePrior
        fiscalYEm2 = endDatePrior + rd(years=-1)
        displayYTD = False
        
    elif endMonth < yearEnd:
        fiscalYE = date(endYear-1, yearEnd, endOfMonth[yearEnd-1])
        fiscalYEm1 = fiscalYE + rd(years=-1)
        fiscalYEm2 = fiscalYE + rd(years=-2)
        displayYTD = True
    
    else:
        fiscalYE = date(endYear, yearEnd, endOfMonth[yearEnd-1])
        fiscalYEm1 = fiscalYE + rd(years=-1)
        fiscalYEm2 = fiscalYE + rd(years=-2)
        displayYTD = True
   
    
    
    relevantDates = {            
            "YTD Start": startDate,
            "YTD End": endDate,
            "YTD Prior Start": startDatePrior,
            "YTD Prior End": endDatePrior,
            "FYE": fiscalYE,
            "FYE-1": fiscalYEm1,
            "FYE-2": fiscalYEm2,
            "Display YTD": displayYTD           
            }
    
    
    return relevantDates['YTD Start']
    
    

    
 
        
    


    
    
    

File: reportBuilder.js
const fs = require('fs');
const moment = require('moment');
const chalk = require('chalk');
const inquirer = require('inquirer');



//input validation//

function initialData () {

    const months = ["January","February","March","April","May","June","July","August","September","October","November","December"]
    const years = ["2014", "2015", "2016", "2017"]
    const endOfMonth = {
        January : 31,
        February: 28,
        March: 31,
        April: 30,
        May: 31,
        June: 30,
        July: 31,
        August: 31,
        September: 30,
        October: 31,
        November: 30,
        December: 30
    }

    const endOfMonthLeap = {
        January : 31,
        February: 29,
        March: 31,
        April: 30,
        May: 31,
        June: 30,
        July: 31,
        August: 31,
        September: 30,
        October: 31,
        November: 30,
        December: 30
    }

    inquirer.prompt([

        {
            name: "firmName",
            type: "input",
            message: chalk.magenta("Please enter the firm name:")
        }, {

            name: "endMonth",
            type: "list",
            message: chalk.greenBright("Please select the period-end month:"),
            choices: months
        },{

            name: "endYear",
            type: "list",
            message: chalk.cyanBright("Please select the period-end year:"),
            choices: years
        }  
    ])
    
    .then(function(answers){

        let firmName = answers.firmName;
        let fullMonthName = answers.endMonth;
        let year = parseInt(answers.endYear);
        let monthNum = months.indexOf(fullMonthName)+1;
        let endDay = endOfMonth[fullMonthName];

        let endDate = moment(`'${year}-${monthNum}-${endDay}'`, 'YYYY-MM-DD').format('YYYY-MM-DD');

        const initialDataObj = {

            firmName: firmName,
            endYear: year,
            endMonth: monthNum,
            endDay: endDay,
            endDate: endDate
        }
     
        console.log(initialDataObj)
        return initialDataObj
        

    })

    }

    function relevantDatesCalc (Year,YTDMonth,FYEMonth = 12) { 
    
        Year = parseInt(Year);
        YTDMonth = parseInt(YTDMonth);

        const endOfMonth = [31,28,31,30,31,30,31,31,30,31,30,31];
        const leapYearEndOfMonth = [31,29,31,30,31,30,31,31,30,31,30,31];

        //end dates of YTD accounting for leap years
        if (Year % 4 == 0) {
                endDate = moment(new Date(Year, YTDMonth-1, leapYearEndOfMonth[YTDMonth-1])).format('YYYY-MM-DD');
                endDatePrior = moment(new Date(Year-1, YTDMonth-1, endOfMonth[YTDMonth-1])).format('YYYY-MM-DD');
        } else if ((Year - 1) % 4 == 0) {
                endDate = moment(new Date(Year, YTDMonth-1, endOfMonth[YTDMonth-1])).format('YYYY-MM-DD');
                endDatePrior = moment(new Date(Year-1, YTDMonth-1, leapYearEndOfMonth[YTDMonth-1])).format('YYYY-MM-DD');
        } else {
                endDate = moment(new Date(Year, YTDMonth-1, endOfMonth[YTDMonth-1])).format('YYYY-MM-DD');
                endDatePrior = moment(new Date(Year-1, YTDMonth-1, endOfMonth[YTDMonth-1])).format('YYYY-MM-DD');          
        }

        // ytd period start dates
        if (FYEMonth === 12) {
                startMonth = 0;
                startYear = Year;
        } else if (YTDMonth < FYEMonth) {
                startMonth = FYEMonth;
                startYear = Year-1;
        } else {
                startMonth = FYEMonth;
                startYear = Year;
        }


        startDate = moment(new Date(startYear, startMonth, 1)).format('YYYY-MM-DD');
        startDatePrior = moment(new Date(startYear-1, startMonth, 1)).format('YYYY-MM-DD');
        

        // fiscal year ends
        if (YTDMonth == FYEMonth ){
            fiscalYE = endDate;
            fiscalYEm1 = endDatePrior;
            fiscalYEm2 = moment(endDatePrior).subtract(1, 'y').format('YYYY-MM-DD');
            displayYTD = false;
            ytdPeriod = 0;

        } else if (YTDMonth < FYEMonth){

            fiscalYE = moment(new Date(Year-1, FYEMonth-1, endOfMonth[FYEMonth-1])).format('YYYY-MM-DD');
            fiscalYEm1 = moment(fiscalYE).subtract(1, 'y').format('YYYY-MM-DD');
            fiscalYEm2 = moment(fiscalYE).subtract(2, 'y').format('YYYY-MM-DD');
            displayYTD = true;
            ytdPeriod = (12 - FYEMonth) + YTDMonth;


        } else {
            fiscalYE = moment (new Date(Year, FYEMonth-1, endOfMonth[FYEMonth-1])).format('YYYY-MM-DD');
            fiscalYEm1 = moment(fiscalYE).subtract(1, 'y').format('YYYY-MM-DD');
            fiscalYEm2 = moment(fiscalYE).subtract(2, 'y').format('YYYY-MM-DD');
            displayYTD = true;
            ytdPeriod = YTDMonth-FYEMonth;


        }


    
    // else:



        const relevantDates =  {

            startDate: startDate,
            endDate: endDate,
            startDatePrior: startDatePrior,
            endDatePrior: endDatePrior,
            fiscalYE: fiscalYE,
            fiscalYEm1: fiscalYEm1,
            fiscalYEm2: fiscalYEm2,
            displayYTD: displayYTD,
            ytdPeriod: ytdPeriod
        }    

       

        console.log(relevantDates);
        return relevantDates;
    }

let Year = process.argv[2];
let YTDMonth = process.argv[3];
let FYEMonth = process.argv[4];

relevantDatesCalc (Year,YTDMonth,FYEMonth) 

    // initialData();